File: shared-components/src/commonTest/kotlin/dev/viethung/components/numberinput/FakeLocaleNumberFormatter.kt
package dev.viethung.components.numberinput

/**
 * Test double for [LocaleNumberFormatter]. Deterministic, no NSNumberFormatter dependency.
 * Knows about en-US, vi-VN, de-DE separators — enough for VM state-machine tests.
 */
class FakeLocaleNumberFormatter : LocaleNumberFormatter {

    private fun separators(locale: String): Pair<String, String> = when (locale) {
        "vi-VN", "de-DE" -> "." to ","
        else -> "," to "."
    }

    override fun format(value: Double, significantDigits: Int, locale: String): String {
        val (groupSep, decSep) = separators(locale)
        val sign = if (value < 0.0) "-" else ""
        val abs = kotlin.math.abs(value)
        val multiplier = pow10(significantDigits)
        val rounded = kotlin.math.round(abs * multiplier).toDouble() / multiplier
        val intPart = rounded.toLong()
        val groupedInt = groupDigits(intPart.toString(), groupSep)
        val fracStr = if (significantDigits == 0) {
            ""
        } else {
            val fracValue = kotlin.math.round((rounded - intPart.toDouble()) * multiplier).toLong()
            decSep + fracValue.toString().padStart(significantDigits, '0')
        }
        return "$sign$groupedInt$fracStr"
    }

    override fun parse(rawText: String, locale: String): Double? {
        if (rawText.isBlank()) return null
        val (groupSep, decSep) = separators(locale)
        return rawText.replace(groupSep, "").replace(decSep, ".").toDoubleOrNull()
    }

    override fun formatLive(rawText: String, locale: String): String {
        val (groupSep, decSep) = separators(locale)
        return liveFormat(rawText, groupSep, decSep) { digits ->
            val n = digits.toLongOrNull() ?: return@liveFormat digits
            groupDigits(n.toString(), groupSep)
        }
    }

    private fun groupDigits(digits: String, groupSep: String): String {
        if (digits.length <= 3) return digits
        val sb = StringBuilder()
        val rev = digits.reversed()
        for (i in rev.indices) {
            if (i > 0 && i % 3 == 0) sb.append(groupSep)
            sb.append(rev[i])
        }
        return sb.reverse().toString()
    }

    private fun pow10(n: Int): Double {
        var result = 1.0
        repeat(n) { result *= 10.0 }
        return result
    }
}

File: swift-package/NumberInputKit/Tests/NumberInputKitTests/FakeLocaleNumberFormatter.swift
import Foundation
@testable import NumberInputKit

/// Test double for `LocaleNumberFormatter`. Deterministic, no `NumberFormatter` dependency.
/// Mirrors `FakeLocaleNumberFormatter.kt` — same separator table, same grouping algorithm.
struct FakeLocaleNumberFormatter: LocaleNumberFormatter {

    private func separators(locale: String) -> (grouping: String, decimal: String) {
        switch locale {
        case "vi-VN", "de-DE": return (".", ",")
        default:                return (",", ".")
        }
    }

    func format(_ value: Double, significantDigits: Int, locale: String) -> String {
        let (groupSep, decSep) = separators(locale: locale)
        let sign = value < 0.0 ? "-" : ""
        let abs = Swift.abs(value)
        let multiplier = pow10(significantDigits)
        let rounded = (abs * multiplier).rounded() / multiplier
        let intPart = Int64(rounded)
        let groupedInt = groupDigits(String(intPart), groupSep: groupSep)
        let fracStr: String
        if significantDigits == 0 {
            fracStr = ""
        } else {
            let fracValue = Int64(((rounded - Double(intPart)) * multiplier).rounded())
            fracStr = decSep + String(fracValue).leftPadded(to: significantDigits)
        }
        return sign + groupedInt + fracStr
    }

    func parse(_ rawText: String, locale: String) -> Double? {
        let trimmed = rawText.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return nil }
        let (groupSep, decSep) = separators(locale: locale)
        let normalised = trimmed
            .replacingOccurrences(of: groupSep, with: "")
            .replacingOccurrences(of: decSep, with: ".")
        return Double(normalised)
    }

    func decimalSeparator(locale: String) -> String { separators(locale: locale).decimal }

    func groupingSeparator(locale: String) -> String { separators(locale: locale).grouping }

    func formatLive(_ rawText: String, locale: String) -> String {
        let (groupSep, decSep) = separators(locale: locale)
        return liveFormat(rawText, groupingSeparator: groupSep, decimalSeparator: decSep) { digits in
            guard let n = Int64(digits) else { return digits }
            return groupDigits(String(n), groupSep: groupSep)
        }
    }

    // MARK: - Helpers

    private func groupDigits(_ digits: String, groupSep: String) -> String {
        guard digits.count > 3 else { return digits }
        var result = ""
        let rev = String(digits.reversed())
        for (i, ch) in rev.enumerated() {
            if i > 0 && i % 3 == 0 { result.append(groupSep) }
            result.append(ch)
        }
        return String(result.reversed())
    }

    private func pow10(_ n: Int) -> Double {
        var r = 1.0
        for _ in 0..<n { r *= 10.0 }
        return r
    }
}

private extension String {
    func leftPadded(to length: Int) -> String {
        guard count < length else { return self }
        return String(repeating: "0", count: length - count) + self
    }
}
